data/blog/2024/02/generic-fuzzy-search-algorithm.md

```diff
@@ -0,0 +1,293 @@
+Let's say you have a list of products, and a user is looking for a "Cozy Sweater." With a fuzzy search algorithm: If they type "Czy Swetr," the fuzzy search considers the similarity and still returns the "Cozy Sweater."
+
+## Different Algorithms
+There are many algorithms out there providing fuzzy search capabilities, some of which you may have encountered. 
+These include the well-known **Levenshtein Distance**, the **Jaro-Winkler** Distance, **N-Gram** and the **Hamming 
+Distance** 
+Algorithm. Additionally, for those in pursuit of speed, two notable options are the **BK Tree** and **Bitmap 
+Algorithm**. It's worth noting that the most efficient and fastest fuzzy search algorithms often include keeping an 
+index of the data, most likely in the form of a trie.
+
+In our example, we use an algorithm that emphasises matching prefixes and will provide a similarity score. The algorithm is fairly fast and easy to understand.
+## Putting search algorithms into practice
+Let's start by setting everything up.
+Go ahead and create a file called `Models`. As the name implies, this is where we define our data models.
+The first one represents a search result:
+```swift
+struct FuzzySearchMatchResult {
+	 let weight: Int
+	 let matchedParts: [NSRange]
+}
+```
+The `weight` is something like a score - the higher, the better.
+And the matchedParts are the ranges of the string that match the search query. Note that we are using an array of ranges not just one range, because there can be a few characters in between the matched characters. Here is an illustration:  
+<figure>
+  <img src="https://github.com/activcoding/codeedit.app/assets/83090745/eaa1c772-a010-4510-b548-1f132a6e367a" />
+  <figcaption>Search results: 'Conviw' query returns ContentView with the highest relevance.</figcaption>
+</figure>
+
+Secondly we need to make sure that the data we searching within is case- and accent-insensitive. This is commonly referred to as 'normalising the string'. To achieve this, we need to add two structs:
+```swift
+struct FuzzySearchCharacter {
+	let content: String
+	let normalisedContent: String
+}
+```
+And
+```swift
+struct FuzzySearchString {
+	var characters: [FuzzySearchCharacter]
+}
+```
+The first one represents one character, the `content` field holds the original character, while `normalisedContent` steps in as its case- and accent-insensitive counterpart.
+And the FuzzySearchString represents a whole word, essentially an array of characters. Here is a representation of the frensh word summer: "été":  
+<figure>
+  <img src="https://github.com/activcoding/codeedit.app/assets/83090745/0ad77727-6bb0-4f2b-a2b4-b1a824c54d11" />
+  <figcaption>Representation of the word 'été' with normalized characters</figcaption>
+</figure>
+
+## Normalise Words
+In order to normalise words, we need to create an extension for Strings. Create a new file called: `String+Normalise.swift`. his naming convention enables others to quickly understand the purpose of the file.
```

Comment:
'This naming convention' - spelling error

Reply:
Thanks for catching that